core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java

```diff
@@ -328,41 +346,119 @@ private void applyOutstandingWalTransactions(
                             // Always set full path when using thread static path
                             operationCompiler.setNowAndFixClock(commitTimestamp);
                             tempPath.of(engine.getConfiguration().getRoot()).concat(tableToken).slash().put(WAL_NAME_BASE).put(walId).slash().put(segmentId);
-                            processWalCommit(writer, tempPath, segmentTxn, operationCompiler, seqTxn);
+                            final long start = microClock.getTicks();
+
+                            if (iTransaction > 0 && transactionMeta.size() < (iTransaction + 2) * TXN_METADATA_LONGS_SIZE) {
+                                // Last few transactions left to process from the list
+                                // of observed transactions built upfront in the beginning of the loop.
+                                // Check if more transaction exist, exit restart the loop to have better picture
+                                // of the future transactions and optimise the application.
+                                if (transactionLogCursor.reset()) {
```

Comment:
nit: this continuation code makes the loop potentially infinite which may lead to other worker starvation. Would be great to introduce some kind of threshold in terms of processed transactions count or time.

Reply:
I'm going to work on fairness as a separate change.
This will involve detecting that there are more tables to apply than the worker threads and terminating this loop earlier.

Reply:
Even when there are no more tables, but the job runs on the shared thread pool, continuing the loop may lead to starvation of other tasks.